controllers/nnf_workflow_controller_helpers.go

```diff
@@ -198,36 +198,81 @@ func (r *NnfWorkflowReconciler) validateStagingDirective(ctx context.Context, wf
+				return nil
+			}
+		}
+		return fmt.Errorf("storage '%s' not found in container profile '%s'", storageName, profile.Name)
+	}
+
+	// Find the wildcard JOB_DW and PERSISTENT_DW arguments and verify that JOB_DW and PERSISTENT_DW
+	// instances exist and that they are found in the container profile
+	var directiveStorages []string // use this list later to validate non-optional storages
+	for arg, storageName := range args {
 		switch arg {
 		case "command", "name", "profile":
 		default:
-			name := val
-			// log.Log.Info("BLAKE: ", "arg", arg, "fsname", name)
-			if strings.HasPrefix(arg, "$JOB_DW_") {
-				if findDirectiveIndexByName(wf, name) == -1 {
-					return nnfv1alpha1.NewWorkflowError(fmt.Sprintf("Job storage instance '%s' not found", name)).WithFatal()
+			if strings.HasPrefix(arg, "JOB_DW_") {
+				if findDirectiveIndexByName(wf, storageName) == -1 {
+					return nnfv1alpha1.NewWorkflowError(fmt.Sprintf("jobdw directive mentioning '%s' not found", storageName)).WithFatal()
+				}
+				if err := checkDirectiveIsInProfile(arg); err != nil {
+					return nnfv1alpha1.NewWorkflowError(err.Error()).WithFatal()
 				}
-			} else if strings.HasPrefix(arg, "$PERSISTENT_DW_") {
-				if err := r.validatePersistentInstanceByName(ctx, name, wf.Namespace); err != nil {
-					return nnfv1alpha1.NewWorkflowError(fmt.Sprintf("Persistent storage instance '%s' not found", name)).WithFatal()
+				directiveStorages = append(directiveStorages, arg)
+			} else if strings.HasPrefix(arg, "PERSISTENT_DW_") {
+				if err := r.validatePersistentInstanceByName(ctx, storageName, wf.Namespace); err != nil {
+					return nnfv1alpha1.NewWorkflowError(fmt.Sprintf("persistent storage instance '%s' not found", storageName)).WithFatal()
 				}
-				if findDirectiveIndexByName(wf, name) == -1 {
-					return nnfv1alpha1.NewWorkflowError(fmt.Sprintf("persistentdw directive mentioning '%s' not found", name)).WithFatal()
+				if findDirectiveIndexByName(wf, storageName) == -1 {
+					return nnfv1alpha1.NewWorkflowError(fmt.Sprintf("persistentdw directive mentioning '%s' not found", storageName)).WithFatal()
 				}
+				if err := checkDirectiveIsInProfile(arg); err != nil {
+					return nnfv1alpha1.NewWorkflowError(err.Error()).WithFatal()
+				}
+				directiveStorages = append(directiveStorages, arg)
 			}
 			// TODO: just ignore anything that doesn't start with JOB_DW/PERSISTENT_DW?
-			// TODO: global lustre?
 		}
 	}
 
-	// TODO: validate profile exists
-	// TODO: validate that supplied storage directives are in the profile storages
+	// Ensure that any storage in the profile that isn't marked as optional is present in the directives
+	checkNonOptionalStorages := func(directiveStorageNames []string) error {
+		for _, storage := range profile.Data.Storages {
+			if !storage.Optional {
+				foundInDirectives := false
+
+				for _, directive := range directiveStorageNames {
+					if storage.Name == directive {
+						foundInDirectives = true
+						break
```

Comment:
You could make this a little simpler by just returning here. Then you wouldn't need the "foundInDirectives" variable.

Reply:
A return here would break out of the outer loop as well, which I don't think we want here. We need to make sure that any non-optional storage is included as an argument in the container directive. Alternatively, we could use a break label.

Or am I missing something here?

Reply:
Nope. You're right. Disregard

Reply:
(this is totally developer preference, so ignore me if you like)

These nested loops with the `found := false` pattern is why I prefer using functions

```go
findInDirectives := func(name string) bool {
  for _, directive := range directiveStorageNames {
    if name == directive {
      return true
    }
    
    return false
}

for _, storage := range profile.Data.Storages {
  if !storage.Optional {    
    if !findInDirectives(storage.name) {
      return fmt.Errorf(...)
    }
  }
}
```

Reply:
I like this. I'll go that route.